feat!: Add support for creating ECS service and container definition

bryantbiggs opened the pull request:
## Description
- Add support for creating cluster CloudWatch log group
- Add support for ECS service connect
- Add support for creating task execution from cluster module; this is for scenarios where you wish to create one task execution role shared across tasks within the cluster (there is also support for individual task execution role at the service level)
- Add support for creating container definition within HCL; this is consumed by the `service` sub-module

## Motivation and Context
- Adds the ability for users to create all ECS resources; now supporting `service` and `container-definition` sub-modules
- Closes #70

## Breaking Changes
- No

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects
- [x] I have executed `pre-commit run -a` on my pull request


⚠️ Note - the changes here have been tested and validated on a number of examples since its used by the [ECS Blueprints](https://github.com/aws-ia/ecs-blueprints/pull/109) project. Once these changes are merged, the ECS Blueprints project will be updated to use this module and [this copy](https://github.com/clowdhaus/terraform-aws-ecs) will be archived/deleted

bryantbiggs commented on modules/service/main.tf:
```diff
@@ -1 +1,1318 @@
+################################################################################
+
+locals {
+  enable_autoscaling = var.create && var.enable_autoscaling
```
No apologies necessary - the more eyes on it, the better! So thank you for taking a look and providing feedback. 

I think thats a great callout - let me look at adding a daemonset example to test and validate this as well. Thank you!

antonbabenko commented on modules/service/variables.tf:
```diff
@@ -0,0 +1,638 @@
+variable "autoscaling_max_capacity" {
+  description = "Maximum number of tasks to run in your service"
+  type        = number
+  default     = 10
```
```suggestion
  default     = null
```

bryantbiggs replied:
See [here](https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76#discussion_r1155118267) - autoscaling is enabled by default since `desired_count` is always ignored

antonbabenko commented on modules/service/variables.tf:
```diff
@@ -0,0 +1,638 @@
+variable "autoscaling_min_capacity" {
+  description = "Minimum number of tasks to run in your service"
+  type        = number
+  default     = 1
```
```suggestion
  default     = null
```

bryantbiggs replied:
Same as above - autoscaling is enabled by default since `desired_count` is always ignored

antonbabenko commented on variables.tf:
```diff
@@ -35,6 +35,34 @@ variable "cluster_settings" {
+variable "cloudwatch_log_group_retention_in_days" {
+  description = "Number of days to retain log events. Default retention - 90 days"
+  type        = number
+  default     = 90
```
```suggestion
  default     = null
```

bryantbiggs replied:
we always provide a default value in our modules to ensure logs are eventually purged and avoid wasting $$ - 90 was selected based on heuristics but we can look at changing the default if necessary. However, we should absolutely avoid `null` here which would retain logs indefinitely

antonbabenko commented on variables.tf:
```diff
@@ -35,6 +35,34 @@ variable "cluster_settings" {
+}
+
+variable "cloudwatch_log_group_retention_in_days" {
+  description = "Number of days to retain log events. Default retention - 90 days"
```
```suggestion
  description = "Number of days to retain log events"
```

bryantbiggs replied:
See above

antonbabenko commented on README.md:
```diff
@@ -2,11 +2,25 @@
 - ECS cluster
 - Fargate capacity providers
 - EC2 AutoScaling Group capacity providers
+- ECS Service sub-module
```
Maybe reduce the number of details shown here next to each feature. It is already described in README of the submodule and examples.

bryantbiggs replied:
done - moved a lot of this stuff to a design doc

antonbabenko reviewed (commented):
I like this PR and the features you've added here.

There are some inline comments and some bigger-scale listed below:
1. All resources should be possible to create separately (and reuse externally created, e.g., cluster, service, task definitions, IAM roles, etc)
2. Do we need to have a submodule for the `service`? OR can we extend the root module and introduce several booleans to create all subresources conditionally? I personally think that having one root module is enough, and it won't be a lot of work to maintain it this way.
3. I think there is no need to have `container-definition` as a separate submodule because it is just a single local map value.

Let me know your thoughts! 🎉

gpdenny commented on modules/service/variables.tf:
```diff
@@ -0,0 +1,638 @@
+
+variable "propagate_tags" {
+  description = "Specifies whether to propagate the tags from the task definition or the service to the tasks. The valid values are `SERVICE` and `TASK_DEFINITION`"
+  type        = bool
```
Should be a string?  :ok_hand:

gpdenny commented on modules/service/main.tf:
```diff
@@ -1 +1,1318 @@
+    delete = try(var.timeouts.delete, null)
+  }
+
+  depends_on = [aws_iam_policy.service]
```
Did some more testing this morning, I think this might need to be depending on the policy attachment, not the policy itself? I managed to get myself into the task draining race condition.

bryantbiggs replied:
good point - I think this is what we do elsewhere as well - updated to depend on policy attachment

victorcechinel commented:
@bryantbiggs any update?

bryantbiggs commented on modules/container-definition/main.tf:
```diff
@@ -0,0 +1,72 @@
+  container_definition = { for k, v in local.definition : k => v if v != null }
+}
+
+resource "aws_cloudwatch_log_group" "this" {
```
There are primarily 2 ways in which logs are created/managed per container definition:
1. Via Cloudwatch logs: When you provide a name to `awslogs-group`, if the CloudWatch log group does not exists, ECS will create it. This means it is created outside of IaC and therefore will not be deleted when the task/container definition is deleted, nor tagged, etc.
2. Via Firelens: Firelens allows users to provide a configuration that will forward logs to various locations, including 3rd party locations. When using Firelens though, users will need to add a FluentBit sidecar to forward logs to the Firelens service which will send logs to the final destiation based on the `firelensConfiguration`

In this module we aim to provide support for both, but with the addition of affording users the ability to manage the Cloudwatch log group through Terraform. This is the similar scenario we face with other modules as well - RDS, Lambda, EKS - that will automatically create log groups when logging is enabled. In our modules we provide support for creating the Cloudwatch log groups so that users can tag the log groups, set the retention period, ensure logs are destroyed when the resources are, etc. This is the reason why the Cloudwatch log resource is provided in this sub-module; to allow users to control the log group through Terraform/IaC

The reason for the the added logic [here](https://github.com/clowdhaus/terraform-aws-ecs/blob/73acc1d808763ccac504eb7ddda7ceef2fc6a1ac/modules/container-definition/main.tf#L6-L16) is to support the following assumptions/defaults:
1. By default, this module enables Cloudwatch logging for the container definition because this is most common for customers. This will create the log group through Terraform to allow the log group to be tagged, a retention period set (save $), and ensure the resource are cleaned up
2. If users wish to use Firelens, the deafult Cloudwatch logging will need to be disabled (this is the reason for `enable_cloudwatch_logging`). An demonstration of this is provided in the examples:
	- [Fluentbit sidecare required for Firelens](https://github.com/clowdhaus/terraform-aws-ecs/blob/73acc1d808763ccac504eb7ddda7ceef2fc6a1ac/examples/fargate/main.tf#L78-L80)
	- [Firelens logging config that depends on fluentbit sidecar](https://github.com/clowdhaus/terraform-aws-ecs/blob/73acc1d808763ccac504eb7ddda7ceef2fc6a1ac/examples/fargate/main.tf#L100-L116)
3. Should, for some reason, users want to enable CloudWatch logging but not control it through Terraform, they can do so by simply disabling the creation of the log group through `create_cloudwatch_log_group = false` 
4. Lastly, if users do not want to enable logging at all, they can disable by setting both `enable_cloudwatch_logging` and `create_cloudwatch_log_group` to `false`

bryantbiggs commented on modules/service/main.tf:
```diff
@@ -1 +1,1318 @@
+}
+
+module "container_definition" {
+  source = "../container-definition"
```
hmm, I'm not sure what the benefit of this would be. What is the concern with having this as a sub-module? The service sub-module is already quite complex - having the container definition as a sub-module reduces its complexity and also provides users a more clear definition of what is supported in the container definition.

bryantbiggs replied:
Let me see if I can get it moved up into the service sub-module

bryantbiggs commented on modules/service/main.tf:
```diff
@@ -1 +1,1318 @@
+  # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-taskset.html
+  count = local.create_task_definition && local.is_external_deployment && !var.ignore_task_definition_changes ? 1 : 0
+
+  service         = try(aws_ecs_service.this[0].id, aws_ecs_service.itd[0].id)
```
what would that look like or what would the benefit be? that would be similar to supporting an externally created Aurora cluster within the Aurora module, or an externally created load balancer in the ALB module. If users have an ECS service, I wouldn't expect them to want or need to use this ECS service module - they would simply deploy their existing service on the ECS cluster

bryantbiggs commented on modules/service/main.tf:
```diff
@@ -1 +1,1318 @@
+  min_capacity = min(var.autoscaling_min_capacity, var.desired_count)
+  max_capacity = max(var.autoscaling_max_capacity, var.desired_count)
+
+  resource_id        = "service/${local.cluster_name}/${try(aws_ecs_service.this[0].name, aws_ecs_service.itd[0].name)}"
```
see comment here https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76#discussion_r1155118267

bryantbiggs commented on modules/service/main.tf:
```diff
@@ -1 +1,1318 @@
+        }
+      }
+
+      scale_in_cooldown  = try(target_tracking_scaling_policy_configuration.value.scale_in_cooldown, 300)
```
Autoscaling is [enabled by default based](https://github.com/clowdhaus/terraform-aws-ecs/blob/a2a9b3d4370f0175dc4441cfc024cfb6989543d5/modules/service/variables.tf#L551-L592) on heuristics. If we set these to `null`, I believe that will [result in the value being 0 ](https://github.com/hashicorp/terraform-provider-aws/blob/d815e303501fcf3943cd95387a6b1302363d2924/internal/service/appautoscaling/policy.go#L197-L208) which would introduce thrashing and service disruption which is not desired

bryantbiggs commented on modules/service/variables.tf:
```diff
@@ -0,0 +1,638 @@
+variable "cpu" {
+  description = "Number of cpu units used by the task. If the `requires_compatibilities` is `FARGATE` this field is required"
+  type        = number
+  default     = 1024
```
See latter part of the variable description - since the default is `FARGATE` (most common), then this value is required so we provide a default value to satisfy

bryantbiggs replied:
its the pattern you typically see with containerized based deployments. the concurrency model shifts from a larger resource with multi-threading and/or parallelism within one pod/container to instead use parallelism across multiple pods/containers. this allows you to scope your resource utilization for "one unit of work" to a pod and then the container orchestrator knows how to scale that based on the workload demand. this allows you to optimize resource utilization efficiently - imagine having pods that were using 4096 for vCPU (4 virtual cores) you have to scale in much larger steps which means you will have periods where you are not fully utilizing the resources provisioned. you also have to factor this into your autoscaling policy - its no longer a 1:1 but a 1:4, your code is more complex if you opt into multi-threading and/or parallelism in order to fully utilize the 4 virtual cores, etc.

bryantbiggs commented on modules/service/variables.tf:
```diff
@@ -0,0 +1,638 @@
+variable "memory" {
+  description = "Amount (in MiB) of memory used by the task. If the `requires_compatibilities` is `FARGATE` this field is required"
+  type        = number
+  default     = 2048
```
Same as above - since the default is `FARGATE` (most common), then this value is required so we provide a default value to satisfy

bryantbiggs replied:
see note above

bryantbiggs commented on modules/service/variables.tf:
```diff
@@ -0,0 +1,638 @@
+variable "requires_compatibilities" {
+  description = "Set of launch types required by the task. The valid values are `EC2` and `FARGATE`"
+  type        = list(string)
+  default     = ["FARGATE"]
```
No - we should have this as the default, it is by far the most common and intentionally setup and configured for Fargate to "just work" out of the box

bryantbiggs commented:
I will get to the rest of the feedback and changes but just wanted to quickly put this here. You typically have a cluster with 1 or more services; within a service you typically have 1 task set or task definition (depending on deployment model); it is possible to have more than one task per service but this is very rare so this PR follows a 1:1 of service to task set/definition, and within a task set or task definition you can have 1-10 container definitions. This is what has driven the 3 different "levels" of modules - the cluster, the service (1 or more per cluster), and the container definition (1-10 per task set/definition). I will take a look at what it would look like to move the container definition into the service, but initially I'm not sure I see the benefit. As its currently written it serves two primary purposes:
1. It simplifies the service module, allowing users to pass 1 or more container definitions 
2. It allows users to create a container definition independently; this separation also allows for clear documentation specific to just the container definition without confusing any of its bits with the service/task

![image](https://user-images.githubusercontent.com/10913471/229296454-c8e0ce54-9d54-4b52-9bf0-ac4540bd8fc2.png)

bryantbiggs commented on modules/container-definition/main.tf:
```diff
@@ -0,0 +1,72 @@
+  retention_in_days = var.cloudwatch_log_group_retention_in_days
+  kms_key_id        = var.cloudwatch_log_group_kms_key_id
+
+  tags = var.tags
```
this is the only AWS resource here - so `var.tags` is essentially tags for the cloudwatch log group

bryantbiggs commented on modules/service/main.tf:
```diff
@@ -1 +1,1318 @@
+  }
+
+  force_delete              = var.force_delete
+  tags                      = var.tags
```
added new variable `var.task_tags` which is merged on both task definition and task set - they serve the same purpose, to create tasks, but are controlled by different means (task set controlled by external deployment controller)

bryantbiggs commented on modules/service/variables.tf:
```diff
@@ -0,0 +1,638 @@
+}
+
+variable "cluster" {
+  description = "ID/ARN of the ECS cluster where the resources will be provisioned"
```
Changed to `cluster_arn`; dropped reference to ID

bryantbiggs commented on main.tf:
```diff
@@ -110,3 +172,156 @@ resource "aws_ecs_capacity_provider" "this" {
+data "aws_iam_policy_document" "task_exec" {
+  count = local.create_task_exec_policy ? 1 : 0
+
+  # Pulled from AmazonECSTaskExecutionRolePolicy
```
we could, yes - but this policy has not changed for several years, users in regulated industries generally want to avoid managed policies since the permissions can change at any time without them knowing, and lastly - it simplifies things to just one policy attached to the role

bryantbiggs commented on modules/container-definition/README.md:
```diff
@@ -0,0 +1,197 @@
+
+Configuration in this directory creates an ECS container definition
+
+By default, the configuration provide creates and uses a CloudWatch log group. To disable creation/use of this behavior you can disable with `enable_cloudwatch_logging` = `false`; for scenarios where using alternate logging methods like FireLens.
```
Updated the phrasing here

bryantbiggs commented on modules/service/main.tf:
```diff
@@ -1 +1,1318 @@
+}
+
+resource "aws_ecs_service" "this" {
+  count = var.create && !var.ignore_task_definition_changes ? 1 : 0
```
see comment above https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76#discussion_r1155117801

bryantbiggs commented on modules/service/main.tf:
```diff
@@ -1 +1,1318 @@
+# Service - Ignore `task_definition`
+################################################################################
+
+resource "aws_ecs_service" "itd" {
```
updated to the longer form - users don't need to type this so you're right, no need to have it short and obfuscated

bryantbiggs commented:
@antonbabenko I made some updates and provided some responses/feedback - please take another look when you have some time and let me know what your thoughts are. I also added a [design doc](https://github.com/clowdhaus/terraform-aws-ecs/blob/master/docs/DESIGN.md) to capture a lot of the information that went into the current proposal. Hopefully that is useful for both the review and future users

antonbabenko commented on .pre-commit-config.yaml:
```diff
@@ -1,6 +1,6 @@
 repos:
   - repo: https://github.com/antonbabenko/pre-commit-terraform
-    rev: v1.77.0
+    rev: v1.77.1
```
Please add this hook after terraform_fmt:
```
      - id: terraform_wrapper_module_for_each
```

antonbabenko reviewed (commented):
Looks very good. The only big thing is related to the submodule for ECS Cluster, and we are good to go!

bryantbiggs commented on README.md:
```diff
@@ -2,11 +2,16 @@
 - EC2 AutoScaling Group capacity providers
+- ECS Service w/ task definition, task set, and container definition support
+
+For more details see the [design doc](https://github.com/terraform-aws-modules/terraform-aws-ecs/blob/master/docs/design.md)
```
```suggestion
For more details see the [design doc](https://github.com/terraform-aws-modules/terraform-aws-ecs/blob/master/docs/DESIGN.md)
```

bryantbiggs commented on README.md:
```diff
@@ -2,11 +2,16 @@
 - EC2 AutoScaling Group capacity providers
+- ECS Service w/ task definition, task set, and container definition support
+
+For more details see the [design doc](https://github.com/terraform-aws-modules/terraform-aws-ecs/blob/master/docs/design.md)
```
fixed in https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76/commits/1db364773f3210fbdfe794eb6ce9b79ec403b482

bryantbiggs commented on modules/container-definition/README.md:
````diff
@@ -0,0 +1,199 @@
+```hcl
+module "fluentbit_ecs_container_definition" {
+  source = "terraform-aws-modules/ecs/aws//modules/container-definition"
+  name = "fluent-bit"
````
fixed in https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76/commits/1db364773f3210fbdfe794eb6ce9b79ec403b482

bryantbiggs commented on docs/DESIGN.md:
```diff
@@ -0,0 +1,289 @@
+   }
+
+   module "ecs_service" {
+     source = "github.com/terraform-aws-modules/terraform-aws-ecs//modules/service"
```
fixed in https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76/commits/1db364773f3210fbdfe794eb6ce9b79ec403b482

bryantbiggs commented on docs/DESIGN.md:
````diff
@@ -0,0 +1,289 @@
+
+   ```hcl
+   module "ecs_service" {
+     source = "github.com/terraform-aws-modules/terraform-aws-ecs//modules/service"
````
fixed in https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76/commits/1db364773f3210fbdfe794eb6ce9b79ec403b482

bryantbiggs commented on docs/DESIGN.md:
````diff
@@ -0,0 +1,289 @@
+
+   ```hcl
+   module "ecs_service" {
+     source = "github.com/terraform-aws-modules/terraform-aws-ecs//modules/service"
````
fixed in https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76/commits/1db364773f3210fbdfe794eb6ce9b79ec403b482

bryantbiggs commented on docs/DESIGN.md:
````diff
@@ -0,0 +1,289 @@
+
+   ```hcl
+   module "ecs_service" {
+     source = "github.com/terraform-aws-modules/terraform-aws-ecs//modules/service"
````
fixed in https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76/commits/1db364773f3210fbdfe794eb6ce9b79ec403b482

bryantbiggs commented on docs/DESIGN.md:
```diff
@@ -0,0 +1,289 @@
+  }
+
+  module "ecs_service" {
+    source = "github.com/terraform-aws-modules/terraform-aws-ecs//modules/service"
```
fixed in https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76/commits/1db364773f3210fbdfe794eb6ce9b79ec403b482

bryantbiggs commented on docs/DESIGN.md:
```diff
@@ -0,0 +1,289 @@
+
+An Amazon ECS service is responsible for managing its associated tasks (via task definition/task set) - the lifecycle of its tasks, the manner in which they are provisioned, their deployment model, network connectivity, etc. As stated previously, Amazon ECS services support one or more task definitions/sets per service. However, this module assumes a service will contain only one task definition/ ...
+
+- When using an external deployment controller, most of the configurations that are normally specified in the service definition are instead specified in the task set. (For users of this project, this is abstractted away and handled by the module).
```
fixed in https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76/commits/1db364773f3210fbdfe794eb6ce9b79ec403b482

antonbabenko reviewed (approved):
LGTM! Great work (as usual)!

antonbabenko reviewed (changes requested):
I hit approved a bit too early :) Let's make a submodule for the ECS cluster.

bryantbiggs commented:
@antonbabenko I've moved the container-definition up to the service module. I added the wrapper pre-commit hook as requested but it seems to be fighting with the format hook 
<img width="857" alt="image" src="https://user-images.githubusercontent.com/10913471/230623714-3f6b1fc3-ffe9-467e-bde2-8c264eb1df59.png">

bryantbiggs commented:
I'm not really digging this influence terragrunt is having over the modules. It seems to be more complexity than value

antonbabenko commented:
Fixed in https://github.com/antonbabenko/pre-commit-terraform/pull/503 , please use `v1.77.2` in this pull-request and rerun it. I couldn't push to your fork:

```
To github.com:terraform-aws-modules/terraform-aws-ecs.git
 ! [remote rejected] clowdhaus/master -> refs/pull/76/head (deny updating a hidden ref)
error: failed to push some refs to 'github.com:terraform-aws-modules/terraform-aws-ecs.git'
```

antonbabenko commented:
> I'm not really digging this influence terragrunt is having over the modules. It seems to be more complexity than value

Terragrunt is one of the tools, but the overall suggestion I have is to have both options:
1. composable ECS units (Cluster, Service, Container-Definition) as submodules
2. root module, which combines calls to Cluster and Service submodules

It will allow users to achieve flexibility as they need regardless of the wrapper tool they use.

---

I see you merged the container-definition submodule into the service. I think it was good to have it as a submodule, though. This way users may use it completely separately from the rest of the ECS resources we manage in this repository.

antonbabenko commented on README.md:
````diff
@@ -2,116 +2,27 @@
 ```hcl
 module "ecs" {
-  source = "terraform-aws-modules/ecs/aws"
+  source = "terraform-aws-modules/ecs"
````
```suggestion
  source = "terraform-aws-modules/ecs/aws"
```

This is in many places throughout the docs.

bryantbiggs replied:
Updated in https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76/commits/c6fef3b1bb52c658ce3cd0a120a38c6e17bef170

antonbabenko commented on main.tf:
```diff
@@ -2,111 +2,183 @@
+  task_exec_iam_role_permissions_boundary = var.task_exec_iam_role_permissions_boundary
+  task_exec_iam_role_tags                 = var.task_exec_iam_role_tags
+  task_exec_iam_role_policies             = var.task_exec_iam_role_policies
+  create_task_exec_policy                 = var.create_task_exec_policy
```
Maybe add a new line before `create_...` and comments like you do in the `service` module to separate blocks.

bryantbiggs replied:
Updated in https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76/commits/c6fef3b1bb52c658ce3cd0a120a38c6e17bef170

antonbabenko commented on modules/cluster/README.md:
```diff
@@ -0,0 +1,214 @@
+# AWS ECS Terraform module
```
```suggestion
# AWS ECS Cluster Terraform module
```

bryantbiggs replied:
Updated in https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/76/commits/c6fef3b1bb52c658ce3cd0a120a38c6e17bef170

antonbabenko commented on outputs.tf:
```diff
@@ -4,35 +4,60 @@
+output "services" {
+  description = "Map of services created and their attributes"
+  value       = module.service
+  sensitive   = true
```
Unfortunately. I really dislike this feature of Terraform... :(

bryantbiggs replied:
Yes! See below (its now removed)

antonbabenko commented on modules/container-definition/outputs.tf:
```diff
@@ -0,0 +1,22 @@
+
+output "container_definition" {
+  description = "Container definition"
+  value       = local.container_definition
```
Should this output be marked as sensitive? I see that in `modules/service/outputs.tf` it is marked so, so here it should probably be also marked?

bryantbiggs replied:
This was my bad - in the examples I was using an SSM parameter data source to pull a container image and the second you use that, it requires outputs to be marked as sensitive. So for now, I removed that and used a static string and will leave this up to users to mark the output as sensitive (if they need to)

antonbabenko reviewed (commented):
Looks very good. It was hard to review in GitHub UI. Minor comments related to docs mostly.

rajibahmed commented:
great work !! I am trying to use this module, really need this to make my billion dollar app work.

antonbabenko reviewed (approved):
👍👍👍

antonbabenko commented:
This PR is included in version 5.0.0 :tada:
